plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/OpenshiftUIModel.java

```diff
@@ -0,0 +1,189 @@
+package org.jboss.tools.openshift.internal.ui.models;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.swt.widgets.Display;
+import org.jboss.tools.openshift.common.core.connection.ConnectionsRegistry;
+import org.jboss.tools.openshift.common.core.connection.ConnectionsRegistrySingleton;
+import org.jboss.tools.openshift.common.core.connection.IConnection;
+import org.jboss.tools.openshift.common.core.connection.IConnectionsRegistryListener;
+import org.jboss.tools.openshift.core.connection.IOpenShiftConnection;
+
+import com.openshift.restclient.model.IResource;
+
+public class OpenshiftUIModel implements IOpenshiftUIElement<IOpenshiftUIElement<?>> {
+	private Map<IOpenShiftConnection, ConnectionWrapper> connections = new HashMap<>();
+	private List<IElementListener> listeners = new ArrayList<IElementListener>();
+
+	private IConnectionsRegistryListener listener;
+	
+	public OpenshiftUIModel() {
+		this(ConnectionsRegistrySingleton.getInstance());
+	}
+
+	public OpenshiftUIModel(ConnectionsRegistry registry) {
+		listener = new IConnectionsRegistryListener() {
+
+			@Override
+			public void connectionRemoved(IConnection connection) {
+				synchronized (connections) {
+					connections.remove(connection);
+				}
+				fireChanged();
+			}
+
+			@Override
+			public void connectionChanged(IConnection c, String property, Object oldValue, Object newValue) {
+				ConnectionWrapper connection = connections.get(c);
+				if (connection == null) {
+					return;
+				}
+				connection.connectionChanged(property, oldValue, newValue);
+			}
+
+			@Override
+			public void connectionAdded(IConnection connection) {
+				if (!(connection instanceof IOpenShiftConnection)) {
+					return;
+				}
+				synchronized (connections) {
+					if (connections.containsKey(connection)) {
+						return;
+					}
+					connections.put((IOpenShiftConnection) connection,
+							new ConnectionWrapper(OpenshiftUIModel.this, (IOpenShiftConnection) connection));
+				}
+				fireChanged();
+			}
+		};
+		Collection<IConnection> allConnections = registry.getAll();

```

Comment:
sync is still needed, but I can get rid of the instanceof, but only if IOpenshiftConnection extends IConnectoin (otherwise the generic types don't compile)